packages/hardhat/contracts/LargeGrant.sol

```diff
@@ -0,0 +1,236 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0 <0.9.0;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+
+abstract contract ERC20 {
+	function transfer(
+		address recipient,
+		uint256 amount
+	) external virtual returns (bool);
+
+	function balanceOf(address account) external view virtual returns (uint256);
+}
+
+contract LargeGrant is AccessControl {
+	bytes32 public constant OWNER_ROLE = keccak256("OWNER_ROLE");
+
+	struct GrantData {
+		address builder;
+		Stage[] stages;
+	}
+
+	struct Stage {
+		uint8 number;
+		Milestone[] milestones;
+	}
+
+	struct Milestone {
+		uint8 number;
+		uint256 amount;
+		bool completed;
+	}
+
+	// used for view functions
+	struct MilestoneData {
+		uint8 stageNumber;
+		uint8 number;
+		uint256 amount;
+		bool completed;
+	}
+
+	mapping(uint256 => GrantData) public grants;
+	address public tokenAddress;
+
+	event AddGrant(
+		uint256 indexed grantId,
+		address indexed to,
+		uint256[] amounts
+	);
+	event AddGrantStage(
+		uint256 indexed grantId,
+		uint8 stageNumber,
+		uint256[] amounts
+	);
+	event CompleteMilestone(
+		uint256 indexed grantId,
+		uint8 stageNumber,
+		uint8 milestoneNumber,
+		uint256 amount
+	);
+	event AddOwner(address indexed newOwner, address indexed addedBy);
+	event RemoveOwner(address indexed removedOwner, address indexed removedBy);
+
+	// Custom errors
+	error GrantAlreadyExists();
+	error GrantDoesNotExist();
+	error MilestoneAlreadyCompleted();
+	error InsufficientContractFunds();
+	error WrongStageNumber();
+	error WrongMilestoneNumber();
+
+	constructor(address _tokenAddress, address[] memory _initialOwners) {
+		tokenAddress = _tokenAddress;
+		_setRoleAdmin(OWNER_ROLE, OWNER_ROLE);
+		for (uint i = 0; i < _initialOwners.length; i++) {
+			_grantRole(OWNER_ROLE, _initialOwners[i]);
+		}
+	}
+
+	function grantData(
+		uint256 _grantId
+	)
+		external
+		view
+		returns (address builder, MilestoneData[] memory milestones)
+	{
+		GrantData storage grant = grants[_grantId];
+		builder = grant.builder;
+
+		uint256 milestonesCount = 0;
+
+		for (uint i = 0; i < grant.stages.length; i++) {
+			milestonesCount += grant.stages[i].milestones.length;
+		}
+
+		milestones = new MilestoneData[](milestonesCount);
+
+		uint256 index = 0;
+		for (uint i = 0; i < grant.stages.length; i++) {
+			for (uint j = 0; j < grant.stages[i].milestones.length; j++) {
+				milestones[index] = MilestoneData({
+					stageNumber: uint8(i + 1),
+					number: grant.stages[i].milestones[j].number,
+					amount: grant.stages[i].milestones[j].amount,
+					completed: grant.stages[i].milestones[j].completed
+				});
+				index++;
+			}
+		}
+	}
+
+	function addGrant(
+		address _builder,
+		uint256 _grantId,
```

Comment:
This will be grantId from web2 db? Maybe lets follow the same notions from Stream contract? 

Like `grantId` correspondence to contract grantId and the `grantNumber` in that GrantData struct represent the grant from db?

Reply:
Yes, the grantId is the large grant id from the web2 db. I thought it was better to just use the same ID.

Reply:
Yup yup I don't see any downside 🙌